### PR TITLE
Add lm_scale to LM rescoring.

### LINKED ISSUE #211: Test results at multiple LM scales, picking best
Can we change the decoding script to test the results with multiple LM scales, say, 0.7 through 1.2 for now?

We already have the property 'lm_scores' on the lattices which dictates the LM part of the scores. (although 'scores' also contains the original acoustic scores plus the LM scores).  
We could clone the 'scores' array, and then repeatedly assign:
  `lats.scores = scores + lats.lm_scores * (lm_scale - 1.0)`
and then get the best paths again.
We will have to check that assigning to the 'scores' property in k2 properly invalidates the forward and backward
scores that are cached in the Fsa, otherwise all output will be the same.

For now let's just always output the WERs from each scale, and also a line that says "best WER" for each test set, so we can easily grep for it.

### PR DESCRIPTION
**Pull request by csukuangfj**
Closes #211 

@danpovey Please try this pull-request.

It requires a pull request I just created in k2 to clear the cache of an FAS if its scores are reassigned.

### CODE REVIEW

File: egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py [hunk 1]
```diff
@@ -34,63 +41,95 @@
 from snowfall.training.mmi_graph import get_phone_symbols
 
 
+def decode_one_batch(batch: Dict[str, Any],
+                     model: AcousticModel,
+                     HLG: k2.Fsa,
+                     output_beam_size: float,
+                     num_paths: int,
+                     use_whole_lattice: bool,
+                     G: Optional[k2.Fsa] = None)->Dict[str, List[List[int]]]:
```
Comment (danpovey) [inline egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py:50]: should probably document this if we intend to check it in.

Comment (csukuangfj) [reply egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py:50]: I've added some documentation to it.

File: egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py [hunk 2]
```diff
@@ -34,63 +41,138 @@
...
+        hyps = get_texts(best_paths, indices)
+        return {'no_rescore': hyps}
+
+    lm_scale_list = [0.7, 0.8, 0.9, 1.0, 1.1, 1.2]
+
+    ans = dict()
+    saved_scores = lattices.scores.clone()
+    for lm_scale in lm_scale_list:
+        lattices.scores = saved_scores.clone()
+        best_paths = decode_with_lm_rescoring(
```
Comment (danpovey) [inline egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py:135]: We could improve the speed here by first doing the composition, and then afterward doing the scaling.

File: egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py [hunk 3]
```diff
@@ -34,63 +41,138 @@
...
+
+@torch.no_grad()
 def decode(dataloader: torch.utils.data.DataLoader, model: AcousticModel,
-           device: Union[str, torch.device], HLG: Fsa, symbols: SymbolTable,
+           HLG: Fsa, symbols: SymbolTable,
            num_paths: int, G: k2.Fsa, use_whole_lattice: bool, output_beam_size: float):
     tot_num_cuts = len(dataloader.dataset.cuts)
     num_cuts = 0
-    results = []  # a list of pair (ref_words, hyp_words)
+    results = defaultdict(list)
```
Comment (danpovey) [inline egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py:154]: can we have a comment about the structure of `results`?

File: egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py (hunk 2 above)
Comment (csukuangfj) [reply egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py:135]: Thanks! I am fixing it.

**Comment by danpovey** [comment]
Thanks!  you can merge when you think it's ready.  Then Desh can have another look at position-dependent phones with this in place.
BTW, in graph.py, when we do
`HLG.lm_scores = HLG.scores.clone()`
.. this might not be 100% correct because these also contain the silence probability at this point; we should perhaps be doing this at the G.fst level.  Of course, this depends on our precise rescoring method; but if we are subtracting the old lm_scores and composing with an n-gram FST, this would wrongly remove the silence probs.

**Comment by aarora8** [comment]
Thank you so much, I will try to use it with position-dependent phones.

**Comment by csukuangfj** [comment]
> we should perhaps be doing this at the G.fst level

Using G.lm_scores can indeed achieve a lower WER than that of HLG.lm_scores.

----

The following screenshots are the results applying `lm_scale` during rescoring.
(The model was trained using only a subset of the training data, i.e., only 100 hours of training data)

# Rescoring with whole lattice (HLG.lm_scores)

Screen Shot 2021-06-15 at 2 36 20 PM

It uses HLG.lm_scores. You can see that lm_scale helps to obtain a better WER 5.63 than the baseline 5.73 (for test-clean).

### Rescoring with whole lattice (G.lm_scores)

Screen Shot 2021-06-15 at 2 36 28 PM

G.lm_scores can further improve WER (5.63 --> 5.56, test-clean)

### Rescoring with n-best list (G.lm_scores, num_paths=100)
Screen Shot 2021-06-15 at 2 39 52 PM

### Rescoring with n-best list (G.lm_scores, num_paths=500)
Screen Shot 2021-06-15 at 2 40 01 PM

File: egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py (hunk 3 above)
Comment (csukuangfj) [reply egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py:158]: More comments are added.

**Comment by danpovey** [comment]
Great!!
You can merge when you think it's ready.
